cmd/add.go

```diff
@@ -227,6 +241,40 @@ func readPaths(file string, splitter bufio.SplitFunc) []string {
 	return paths
 }
 
+// categorisePaths categorises each path in a given slice into either a
+// directory or file and appends the path to the corresponding slice.
+func categorisePaths(filesAndDirs, files, dirs []string) ([]string, []string) {
+	dirSet := make(map[string]bool)
+
+	for _, path := range filesAndDirs {
+		if pathIsDir(path) {
+			dirs = append(dirs, path)
+			dirSet[path] = true
+		}
+	}
+
+	for _, path := range filesAndDirs {
+		if !pathIsDir(path) {
+			if !dirSet[filepath.Dir(path)] {
```

Comment:
These two `if` statements can be combined. It make also be better to just check the `dirSet` map for the path instead of re-statting it.